File: bnet/attractors.jl
struct BNet
    number_of_nodes::Int
    weights::Array{Float64,2}
    thresholds::Vector{Float64}
    BNet(ws, ts) = new(size(ws,1), ws, ts)
end

load(::Type{BNet}, wfile, tfile) = BNet(readdlm(wfile), vec(readdlm(tfile)))

Base.size(net) = net.number_of_nodes

volume(net) = 1 << size(net)

function fire!(net, state, next_state)
    temp = net.weights * state
    for i in 1:length(state)
        next_state[i] = if temp[i] > net.thresholds[i]
            1
        elseif temp[i] < net.thresholds[i]
            0
        else
            state[i]
        end
    end
    next_state
end

fire(net, state) = fire!(net, state, similar(state))

function encode(state)
    const N = length(state)
    code = 0
    for i in 1:N
        code = 2code + state[N - i + 1]
    end
    code + 1
end

function transitions(net)
    const N, V = size(net), volume(net)
    trans = Array{Int}(V)
    state = zeros(Int, N)
    next_state = fire(net, state)

    code, next_code = 1, encode(next_state)
    trans[code] = next_code

    while code < V
        for i in 1:N
            if state[i] == 0
                state[i] += 1
                state[1:i-1] = 0
                break
            end
        end
        code += 1

        fire!(net, state, next_state)
        next_code = encode(next_state)
        trans[code] = next_code
    end

    trans
end

function attractors(net)
    const V = volume(net)
    const trans = transitions(net)
    seen = zeros(Bool, V)
    basins = fill(-1, V)
    basin_number = 0

    attractors = Vector{Int}[]
    state_stack = Int[]
    cycle = Int[]
    for initial_state in trans
        if seen[initial_state]
            continue
        end

        state = initial_state
        in_cycle = false
        terminus = next_state = trans[state]
        seen[state] = true

        while !seen[next_state]
            push!(state_stack, state)
            state = next_state
            terminus = next_state = trans[state]
            seen[state] = true
        end

        basin = basins[next_state]
        if basins[next_state] == -1
            basin = basin_number
            basin_number += 1
            push!(cycle, state)
            in_cycle = (terminus != state)
        end

        basins[state] = basin
        while !isempty(state_stack)
            state = pop!(state_stack)
            basins[state] = basin
            if in_cycle
                push!(cycle, state)
                in_cycle = (terminus != state)
            end
        end

        if !isempty(cycle)
            push!(attractors, cycle)
            cycle = Int[]
        end
    end
    attractors
end

function main(wfile, tfile)
    const net = load(BNet, wfile, tfile)

    duration = @elapsed begin
        attrs = attractors(net)
    end
    println("First Invocation:  $(duration)s\n")
    
    duration = @elapsed begin
        attrs = attractors(net)
    end
    w = ceil(log2(length(attrs)))
    for (i, attr) in enumerate(attrs)
        print("Attractor $i: ")
        for state in attr
            print("$(state-1) ")
        end
        println()
    end
    println("Second Invocation: $(duration)s")
end

main(ARGS[1], ARGS[2])

File: bnet/C++/attractors.cpp
#include <chrono>
#include <cmath>
#include <fstream>
#include <iomanip>
#include <iostream>
#include <iterator>
#include <numeric>
#include <stack>
#include <vector>

auto operator<<(std::ostream &out, std::vector<int> const& v) -> std::ostream&
{
    std::copy(std::begin(v), std::end(v), std::ostream_iterator<int>(out, " "));
    return out;
}

auto encode(std::vector<int> const &state)
{
    auto const N = std::size(state);
    auto code = 0ull;
    for (size_t i = 0; i < N; ++i)
    {
        code = 2*code + state.at(N - i - 1);
    }
    return code;
}

auto count_lines(char const *filename)
{
    auto number_of_lines = 0ull;
    std::string line;
    auto handle = std::ifstream{filename};
    while (std::getline(handle, line))
    {
        number_of_lines += 1;
    }
    return number_of_lines;
}

struct BNet
{
    size_t number_of_nodes;
    std::vector<std::vector<double>> weights;
    std::vector<double> thresholds;

    BNet(size_t n)
    {
        this->number_of_nodes = n;
        this->weights =
            std::vector<std::vector<double>>(n, std::vector<double>(n));
        this->thresholds = std::vector<double>(n);
    }

    BNet(BNet const&) = delete;
    BNet(BNet&&) = default;

    auto operator=(BNet const&) -> BNet& = delete;
    auto operator=(BNet&&) -> BNet& = default;

    auto size() const
    {
        return number_of_nodes;
    }

    auto volume() const
    {
        return static_cast<size_t>(1 << number_of_nodes);
    }

    static auto read(char const *wfile, char const *tfile)
    {
        auto number_of_nodes = count_lines(wfile);

        auto net = BNet{number_of_nodes};

        {
            auto handle = std::ifstream{wfile};
            for (auto &row : net.weights)
            {
                for (auto &weight : row)
                {
                    handle >> weight;
                }
            }
            handle.close();
        }

        {
            auto handle = std::ifstream{tfile};
            for (auto &threshold : net.thresholds)
            {
                handle >> threshold;
            }
            handle.close();
        }

        return net;
    }

    auto fire(std::vector<int> const &state, std::vector<int> &next_state) const
    {
        auto temp = std::vector<double>(std::size(state), 0.0);
        for (auto i = 0ull; i < std::size(state); ++i)
        {
            auto const &row = weights.at(i);
            temp.at(i) = std::inner_product(std::begin(row), std::end(row),
                    std::begin(state), 0.0);
        }
        for (auto i = 0ull; i < std::size(state); ++i)
        {
            if (temp.at(i) > thresholds.at(i))
            {
                next_state.at(i) = 1;
            }
            else if (temp.at(i) < thresholds.at(i))
            {
                next_state.at(i) = 0;
            }
            else
            {
                next_state.at(i) = state.at(i);
            }
        }

        return next_state;
    }

    auto fire(std::vector<int> const &state) const 
    {
        auto next_state = std::vector<int>(std::size(state));
        fire(state, next_state);
        return next_state;
    }

    auto transitions() const
    {
        auto const N = size();
        auto const V = volume();

        auto trans = std::vector<size_t>(V, -1);
        auto state = std::vector<int>(N, 0);
        auto next_state = fire(state);

        auto code = 0ull;
        auto next_code = encode(next_state);
        trans.at(code) = next_code;

        while (code < V - 1)
        {
            for (auto i = 0ull; i < N; ++i)
            {
                if (state.at(i) == 0)
                {
                    ++state.at(i);
                    for (auto j = 0ull; j < i; ++j)
                    {
                        state.at(j) = 0;
                    }
                    break;
                }
            }
            ++code;

            fire(state, next_state);
            next_code = encode(next_state);
            trans.at(code) = next_code;
        }

        return trans;
    }

    auto attractors() const
    {
        size_t const V = volume();

        auto const trans = transitions();

        auto seen = std::vector<bool>(V, false);
        auto basins = std::vector<int>(V, -1);
        auto basin_number = 0;

        auto attractors = std::vector<std::vector<int>>();
        auto state_stack = std::stack<std::size_t>();
        auto cycle = std::vector<int>();
        for (auto initial_state : trans)
        {
            if (seen.at(initial_state))
            {
                continue;
            }

            auto state = initial_state;
            auto in_cycle = false;
            auto next_state = trans.at(state);
            auto terminus = next_state;

            seen.at(state) = true;
            while (!seen.at(next_state))
            {
                state_stack.push(state);
                state = next_state;
                terminus = next_state = trans.at(state);
                seen.at(state) = true;
            }

            auto basin = basins.at(next_state);
            if (basins.at(next_state) == -1)
            {
                basin = basin_number;
                ++basin_number;
                cycle.push_back(state);
                in_cycle = (terminus != state);
            }

            basins.at(state) = basin;
            while (!state_stack.empty())
            {
                state = state_stack.top();
                state_stack.pop();

                basins.at(state) = basin;

                if (in_cycle)
                {
                    cycle.push_back(state);
                    in_cycle = (terminus != state);
                }
            }

            if (!cycle.empty())
            {
                attractors.push_back(cycle);
                cycle = std::vector<int>();
            }
        }

        return attractors;
    }
};

auto operator<<(std::ostream& out, BNet const& net) -> std::ostream&
{
    out << "number of nodes: " << std::size(net) << std::endl;
    out << "weights:" << std::endl;
    for (auto const &row : net.weights)
    {
        out << "  ";
        for (auto const &weight : row)
        {
            out << std::setw(2) << std::right;
            out << weight << '\t';
        }
        out << std::endl;
    }
    out << "thresholds:" << std::endl;
    for (auto const &threshold : net.thresholds)
    {
        out << std::setw(4) << std::setprecision(1) << std::right;
        out << threshold << '\t';
    }
    return out;
}

auto main(int argc, char **argv) -> int
{
    using namespace std::chrono;

    if (argc < 3)
    {
        std::cerr << "usage: attractors <weights> <thresholds>" << std::endl;
        return 1;
    }

    auto net = BNet::read(argv[1], argv[2]);

    auto start = steady_clock::now();
    auto attrs = net.attractors();
    auto stop = steady_clock::now();
    auto duration = 1e-9 * duration_cast<nanoseconds>(stop - start).count();
    std::cout << "First Invocation:  " << duration << "s\n" << std::endl;

    start = steady_clock::now();
    attrs = net.attractors();
    stop = steady_clock::now();
    duration = 1e-9 * duration_cast<nanoseconds>(stop - start).count();
    for (auto i = 0ull; i < std::size(attrs); ++i)
    {
        std::cout << "Attractor " << i << ": " << attrs.at(i) << std::endl;
    }
    std::cout << "Second Invocation: " << duration << "s" << std::endl;
}
